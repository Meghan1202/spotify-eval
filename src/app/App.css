.App {
  text-align: center;
  display: flex;
  flex-direction: column;
  background-color: #f5f5f5;
}

@media (prefers-reduced-motion: no-preference) {
  .App-logo {
    animation: App-logo-spin infinite 20s linear;
  }
}

.Records-Container {
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
}

.App-header {
  background-color: #282c34;
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
  color: white;
}

.App-link {
  color: #61dafb;
}

