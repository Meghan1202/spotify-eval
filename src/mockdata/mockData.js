const mockLike = { like: false, count: 1 };

const mockSongs = [{
  id: '1',
  name: 'song_name',
  genre: { id: '1', name: 'genre_name' },
  artist: { id: '1', name: 'abc' },
  albumArtUrl: 'abc.com',
  publishedAt: '',
  likes: 1,
  isLikedAlready: true,
}];

module.exports = { mockLike, mockSongs };
