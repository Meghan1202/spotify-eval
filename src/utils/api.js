const axios = require('axios');

const token = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJqdGkiOiIxN2I2NjI4ZC0xZWJkLTRjMGItOGU3ZS0yZTFlYTgwYzc3MzEiLCJuYW1lIjoiTWVnaGFuYSIsInJvbGUiOjAsInN1YmplY3QiOiJUZWNoIFVuaXYgMjAyMSIsImVhc3Rlcl9lZ2ciOiJHb29kIGx1Y2shIiwiYXVkIjoibXVzaWNyZWNvcmRzLnRlY2h1bml2LmNvbSIsIm5iZiI6MTYxNTc4NDQzNCwiZXhwIjoxNjE4Mzc2NDM0LCJpYXQiOjE2MTU3ODQ0MzQsImlzcyI6Ik11c2ljIFJlY29yZHMifQ.zndapgDdsHymhHBydkTyQcrA6rmhSEJtoJ3PtGQVoQs';
const apiBaseURL = 'https://musicrecordsapi.azurewebsites.net';

const apiInstance = axios.create({
  baseURL: apiBaseURL,
  timeout: 10000,
  headers: { Authorization: `Bearer ${token}` },
});

const getLikesDataForSong = async (id) => {
  const likesResponse = await apiInstance.get(`/api/records/${id}/likes`);
  const likesData = await likesResponse.data;
  return likesData.data;
};

const getSongsData = async () => {
  const songsResponse = await apiInstance.get('/api/records');
  const songsData = await Promise.all(songsResponse.data.data.map(async (song) => {
    const likesDataForRecord = await getLikesDataForSong(song.id);
    return {
      ...song,
      likes: likesDataForRecord.count,
      isLikedAlready: likesDataForRecord.like,
    };
  }));
  return songsData;
};

const getRecordsBasedOnGenres = async () => {
  const genreData = {};
  const songs = await getSongsData();
  songs.forEach((song) => {
    if (genreData[song.genre.name]) {
      genreData[song.genre.name] = [...genreData[song.genre.name], song];
    } else {
      genreData[song.genre.name] = [song];
    }
  });
  return genreData;
};

const patchSongLikeData = async (id, currentState) => {
  const likePatchResponse = await apiInstance.patch(`/api/records/${id}/likes`, {
    like: !currentState,
  });
  const likePatchData = await likePatchResponse.data;
  return likePatchData;
};

module.exports = {
  getSongsData,
  getLikesDataForSong,
  patchSongLikeData,
  getRecordsBasedOnGenres,
};
