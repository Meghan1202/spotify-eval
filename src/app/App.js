import './App.css';
import React, { useState } from 'react';
import { Route, Switch } from 'react-router-dom';
import NavBar from '../components/NavBar/NavBar';
import api from '../utils/api';
import AllSongs from '../pages/AllSongs/AllSongs';
import LandingPage from '../pages/LandingPage/LandingPage';
import GenreViewPage from '../pages/GenreViewPage/GenreViewPage';

function App() {
  const [songs, setSongs] = useState([]);
  const [genreData, setGenreData] = useState({});

  const updateSongs = (id, currentState) => {
    const songsModified = songs.map((song) => {
      if (song.id === id) {
        return {
          ...song,
          likes: currentState ? song.likes + 1 : song.likes - 1,
          isLikedAlready: currentState,
        };
      }
      return song;
    });
    const genres = api.getRecordsBasedOnGenres(songsModified);
    const genreDataModified = { ...genres };
    setSongs(songsModified);
    setGenreData(genreDataModified);
  };

  const getGenre = async () => {
    const genres = await api.getRecordsBasedOnGenres(songs);
    setGenreData(genres);
  };

  const getSongs = async () => {
    const songsData = await api.getSongsData();
    setSongs(songsData);
  };

  return (
    <div className="App">
      <NavBar />
      {
        songs.length ? (
          <Route path="/allSongs">
            <AllSongs
              songs={songs}
              toggleView={getGenre}
              patchLike={api.patchSongLikeData}
              updateSongs={updateSongs}
            />
          </Route>
        ) : <Route path="/"><LandingPage getSongs={getSongs} /></Route>
      }
      <Switch>
        <Route path="/genre"><GenreViewPage genreData={genreData} patchLike={api.patchSongLikeData} updateSongs={updateSongs} toggleView={getSongs} /></Route>
      </Switch>
    </div>
  );
}

export default App;
