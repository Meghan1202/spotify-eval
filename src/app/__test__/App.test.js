import {
  render, waitFor,
} from '@testing-library/react';
import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import App from '../App';
import api from '../../utils/api';
import { mockGenre, mockSongs } from '../../mockdata/mockData';

describe(App.name, () => {
  beforeEach(() => {
    jest.spyOn(api, 'getSongsData').mockResolvedValueOnce(mockSongs).mockResolvedValueOnce(null);
    jest.spyOn(api, 'getRecordsBasedOnGenres').mockResolvedValue(mockGenre).mockResolvedValueOnce(null);
  });
  afterEach(() => {
    jest.clearAllMocks();
  });
  test('should match the snapshot', async () => {
    const { container } = await waitFor(() => render(<BrowserRouter><App /></BrowserRouter>));
    expect(container).toMatchSnapshot();
  });
  test('should route to GenreViewPage', async () => {
    await waitFor(() => render(<BrowserRouter><App /></BrowserRouter>));
  });
});
