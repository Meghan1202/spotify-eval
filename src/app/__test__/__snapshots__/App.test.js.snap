// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match the snapshot 1`] = `
<div>
  <div
    class="App"
  >
    <div
      class="Nav-Bar-Container"
    >
      My  
      <strong>
         Record 
      </strong>
        Shelf
    </div>
    <div
      class="Landing-Body"
    >
      <div
        class="Landing-Page"
      >
        <p>
          :(
           
          <br />
          Seems a bit empty in here...
          <br />
        </p>
        <a
          class="Sync-Buton"
          href="/allSongs"
        >
          Sync
        </a>
      </div>
    </div>
  </div>
</div>
`;
